File: src/application/useCases/managePeople.ts
import {
  checkSettlement,
  computePersonBalance,
  remainingOfObligation,
  type PersonBalance,
} from '../../domain/ledger'
import { formatMoney } from '../../domain/formatMoney'
import type { Halalas } from '../../domain/money'
import type {
  Id,
  Obligation,
  ObligationKind,
  Person,
  PersonAllocation,
  Settlement,
} from '../../domain/entities/types'
import type {
  AllocationRepository,
  Clock,
  IdGenerator,
  ObligationRepository,
  PersonRepository,
  SettlementRepository,
  TransactionRepository,
  UnitOfWork,
} from '../ports/repositories'

/**
 * ManagePeople — الأشخاص والديون والأمانات.
 *
 * منطق الحساب كله في `domain/ledger.ts` ومختبَر بـ25 حالة من `spec/06`.
 * هذا الملف يربطه بالتخزين ولا يعيد حسابًا.
 *
 * `spec/02`: «اعرض منفصلين: **لك عنده** و**له عندك**… سجّل نوع الالتزام
 *  داخليًا منفصلًا للديون والأمانات… **لا تقاص تلقائي** بين لك وعليك.»
 */

export class PeopleError extends Error {}

export interface PersonRow {
  person: Person
  balance: PersonBalance
  obligations: { obligation: Obligation; remainingMinor: Halalas }[]
}

export interface ManagePeopleDeps {
  people: PersonRepository
  obligations: ObligationRepository
  settlements: SettlementRepository
  allocations: AllocationRepository
  txns: TransactionRepository
  uow: UnitOfWork
  ids: IdGenerator
  clock: Clock
}

export function makeManagePeople(deps: ManagePeopleDeps) {
  /** كل الأشخاص بأرصدتهم. القراءة محدودة بعدد الأشخاص لا بالعمليات. */
  async function listWithBalances(): Promise<PersonRow[]> {
    const people = await deps.people.listAll()
    const rows: PersonRow[] = []

    for (const person of people) {
      const obligations = await deps.obligations.listByPerson(person.id)
      const settlements = await deps.settlements.listByObligations(obligations.map((o) => o.id))
      rows.push({
        person,
        balance: computePersonBalance(person.id, obligations, settlements),
        obligations: obligations
          .map((obligation) => ({
            obligation,
            remainingMinor: remainingOfObligation(obligation, settlements),
          }))
          // المسدَّد بالكامل لا يُعرض في القائمة النشطة
          .filter((o) => o.remainingMinor > 0),
      })
    }

    // الأكثر رصيدًا أولًا، والمؤرشف أخيرًا
    return rows.sort((a, b) => {
      if (a.person.archived !== b.person.archived) return a.person.archived ? 1 : -1
      const total = (r: PersonRow) =>
        r.balance.receivableMinor + r.balance.payableLoanMinor + r.balance.payableCustodyMinor
      return total(b) - total(a)
    })
  }

  async function addPerson(name: string): Promise<Person> {
    const trimmed = name.trim()
    if (!trimmed) throw new PeopleError('اكتب اسم الشخص')
    // spec/04: حد اسم الشخص 80 حرفًا، معلن ومتحقَّق منه
    if (trimmed.length > 80) throw new PeopleError('الاسم أطول من 80 حرف')

    const existing = await deps.people.listAll()
    if (existing.some((p) => p.name.trim() === trimmed)) {
      throw new PeopleError(`فيه شخص اسمه «${trimmed}» موجود قبل كده`)
    }

    const person: Person = { id: deps.ids.next('person'), name: trimmed, archived: false }
    await deps.people.save(person)
    return person
  }

  /**
   * أرشفة شخص. **لا حذف** — `spec/03`: «لا حذف للحساب ذي سجل».
   * المؤرشف يختفي من الاختيار ويبقى سجله ورصيده ظاهرين.
   */
  async function archivePerson(personId: Id, archived: boolean): Promise<void> {
    const people = await deps.people.listAll()
    const person = people.find((p) => p.id === personId)
    if (!person) throw new PeopleError('الشخص ده مش موجود')
    await deps.people.save({ ...person, archived })
  }

  /**
   * ينشئ التزامًا من عملية: دفعت عن حد، أو سلّفت، أو استلمت قرضًا/أمانة.
   *
   * `spec/02`: «مجموع التخصيصات للشخص **لا يتجاوز قيمة الشراء**».
   */
  async function linkToPerson(input: {
    transactionId: Id
    personId: Id
    kind: ObligationKind
    amountMinor: Halalas
    /** «هدية» لا تُنشئ التزامًا وتُبقي المصروف كاملًا (spec/06). */
    asGift?: boolean
  }): Promise<{ obligation: Obligation | null; allocation: PersonAllocation }> {
    if (input.amountMinor <= 0) throw new PeopleError('المبلغ لازم يكون أكبر من صفر')

    const [transaction] = await deps.txns.findByIds([input.transactionId])
    if (!transaction) throw new PeopleError('العملية دي مش موجودة')

    const existing = await deps.allocations.listByTransactionIds([input.transactionId])
    const already = existing.reduce((sum, a) => sum + a.amountMinor, 0)
    if (already + input.amountMinor > transaction.amountMinor) {
      throw new PeopleError(
        `مجموع التخصيصات (${formatMoney(already + input.amountMinor)}) أكبر من قيمة ` +
          `العملية (${formatMoney(transaction.amountMinor)})`,
      )
    }

    return deps.uow.run(async () => {
      const allocation: PersonAllocation = {
        id: deps.ids.next('alloc'),
        transactionId: input.transactionId,
        personId: input.personId,
        allocationKind: input.asGift ? 'gift' : 'receivable',
        amountMinor: input.amountMinor,
        currency: transaction.currency,
      }
      await deps.allocations.saveMany([allocation])

      // الهدية تخصيص بلا التزام — spec/06: «لا التزام عليه»
      if (input.asGift) return { obligation: null, allocation }

      const obligation: Obligation = {
        id: deps.ids.next('obl'),
        personId: input.personId,
        originTransactionId: input.transactionId,
        kind: input.kind,
        originalMinor: input.amountMinor,
        currency: transaction.currency,
      }
      await deps.obligations.saveMany([obligation])
      return { obligation, allocation }
    })
  }

  /**
   * يسوّي التزامًا. يمر على `checkSettlement` أولًا،
   * فالزيادة **تُرفض بتفسير** ولا تُبتلع ولا تُسحب من التزام آخر (spec/06).
   */
  async function settle(input: {
    obligationId: Id
    personId: Id
    amountMinor: Halalas
    /** العملية التي تمثل السداد، إن وُجدت. */
    transactionId?: Id
  }): Promise<Settlement> {
    const obligations = await deps.obligations.listByPerson(input.personId)
    const obligation = obligations.find((o) => o.id === input.obligationId)
    if (!obligation) throw new PeopleError('الالتزام ده مش موجود')

    const settlements = await deps.settlements.listByObligations([obligation.id])
    const check = checkSettlement(obligation, settlements, input.amountMinor)
    if (!check.allowed) throw new PeopleError(check.reason ?? 'التسوية مرفوضة')

    const settlement: Settlement = {
      id: deps.ids.next('stl'),
      transactionId: input.transactionId ?? '',
      obligationId: obligation.id,
      amountMinor: check.settledMinor,
    }
    await deps.settlements.saveMany([settlement])
    return settlement
  }

  return { listWithBalances, addPerson, archivePerson, linkToPerson, settle }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/ManagePeople.kt
package app.masroufy.usecase

import app.masroufy.core.AllocationKind
import app.masroufy.core.Currency
import app.masroufy.core.Halalas
import app.masroufy.core.Id
import app.masroufy.core.MAX_SAFE_HALALAS
import app.masroufy.core.Obligation
import app.masroufy.core.ObligationKind
import app.masroufy.core.Person
import app.masroufy.core.PersonAllocation
import app.masroufy.core.PersonBalance
import app.masroufy.core.Settlement
import app.masroufy.core.computePersonBalance
import app.masroufy.core.formatMoney
import app.masroufy.core.jsTrim
import app.masroufy.core.remainingOfObligation
import app.masroufy.port.AllocationRepository
import app.masroufy.port.Clock
import app.masroufy.port.IdGenerator
import app.masroufy.port.ObligationRepository
import app.masroufy.port.PersonRepository
import app.masroufy.port.SettlementRepository
import app.masroufy.port.SettlementWriter
import app.masroufy.port.TransactionRepository
import app.masroufy.port.UnitOfWork

/**
 * ManagePeople — نقل `managePeople.ts`: الأشخاص والديون والأمانات.
 * منطق الحساب كله في `core/Ledger.kt` — الملف ده بيربطه بالتخزين وما بيعيدش حساب.
 * spec/02: «اعرض منفصلين: لك عنده وله عندك… **لا تقاص تلقائي** بين لك وعليك».
 */

data class PersonObligationRow(val obligation: Obligation, val remainingMinor: Halalas)

data class PersonRow(
    val person: Person,
    val balance: PersonBalance,
    val obligations: List<PersonObligationRow>,
)

data class ManagePeopleDeps(
    val people: PersonRepository,
    val obligations: ObligationRepository,
    val settlements: SettlementRepository,
    val settlementWriter: SettlementWriter,
    val allocations: AllocationRepository,
    val txns: TransactionRepository,
    val uow: UnitOfWork,
    val ids: IdGenerator,
    val clock: Clock,
)

data class LinkResult(val obligation: Obligation?, val allocation: PersonAllocation)

private val REQUEST_ID = Regex("^[a-zA-Z0-9_-]{1,100}$")

class ManagePeople(private val deps: ManagePeopleDeps) {
    /** كل الأشخاص بأرصدتهم — القراية محدودة بعدد الأشخاص مش بالعمليات. */
    suspend fun listWithBalances(): List<PersonRow> {
        val rows = deps.people.listAll().map { person ->
            val obligations = deps.obligations.listByPerson(person.id)
            val settlements = deps.settlements.listByObligations(obligations.map { it.id })
            PersonRow(
                person = person,
                balance = computePersonBalance(person.id, obligations, settlements),
                obligations = obligations
                    .map { PersonObligationRow(it, remainingOfObligation(it, settlements)) }
                    // المسدَّد بالكامل ما بيتعرضش في القايمة النشطة
                    .filter { it.remainingMinor > 0 },
            )
        }

        // الأكتر رصيدًا الأول، والمؤرشف آخر واحد
        fun total(r: PersonRow) = r.balance.receivableMinor + r.balance.payableLoanMinor + r.balance.payableCustodyMinor
        return rows.sortedWith(compareBy<PersonRow> { it.person.archived }.thenByDescending { total(it) })
    }

    suspend fun addPerson(name: String): Person {
        val trimmed = jsTrim(name)
        if (trimmed.isEmpty()) throw IllegalStateException("اكتب اسم الشخص")
        // spec/04: حد اسم الشخص 80 حرف، معلن ومتحقَّق منه
        if (trimmed.length > 80) throw IllegalStateException("الاسم أطول من 80 حرف")

        if (deps.people.listAll().any { jsTrim(it.name) == trimmed }) {
            throw IllegalStateException("فيه شخص اسمه «$trimmed» موجود قبل كده")
        }

        val person = Person(id = deps.ids.next("person"), name = trimmed, archived = false)
        deps.people.save(person)
        return person
    }

    /** أرشفة — **مفيش حذف** (spec/03: «لا حذف للحساب ذي سجل»). */
    suspend fun archivePerson(personId: Id, archived: Boolean) {
        val person = deps.people.listAll().find { it.id == personId } ?: throw IllegalStateException("الشخص ده مش موجود")
        deps.people.save(person.copy(archived = archived))
    }

    /**
     * بينشئ التزامًا من عملية: دفعت عن حد أو سلّفت أو استلمت قرض/أمانة.
     * spec/02: «مجموع التخصيصات للشخص **لا يتجاوز قيمة الشراء**»،
     * والهدية تخصيص بلا التزام (spec/06).
     */
    suspend fun linkToPerson(
        transactionId: Id,
        personId: Id,
        kind: ObligationKind,
        amountMinor: Halalas,
        asGift: Boolean = false,
    ): LinkResult {
        if (amountMinor <= 0) throw IllegalStateException("المبلغ لازم يكون أكبر من صفر")

        val transaction = deps.txns.findByIds(listOf(transactionId)).firstOrNull()
            ?: throw IllegalStateException("العملية دي مش موجودة")

        val already = deps.allocations.listByTransactionIds(listOf(transactionId)).fold(0L) { sum, a -> sum + a.amountMinor }
        if (already + amountMinor > transaction.amountMinor) {
            throw IllegalStateException(
                "مجموع التخصيصات (${formatMoney(already + amountMinor)}) أكبر من قيمة العملية (${formatMoney(transaction.amountMinor)})",
            )
        }

        return deps.uow.run {
            val allocation = PersonAllocation(
                id = deps.ids.next("alloc"),
                transactionId = transactionId,
                personId = personId,
                allocationKind = if (asGift) AllocationKind.GIFT else AllocationKind.RECEIVABLE,
                amountMinor = amountMinor,
                currency = transaction.currency,
            )
            deps.allocations.saveMany(listOf(allocation))

            if (asGift) return@run LinkResult(obligation = null, allocation = allocation)

            val obligation = Obligation(
                id = deps.ids.next("obl"),
                personId = personId,
                originTransactionId = transactionId,
                kind = kind,
                originalMinor = amountMinor,
                currency = transaction.currency,
            )
            deps.obligations.saveMany(listOf(obligation))
            LinkResult(obligation = obligation, allocation = allocation)
        }
    }

    /** بيسوّي التزامًا — الزيادة بتترفض بتفسير، ونفس معرّف الطلب مرتين = نفس التسوية. */
    suspend fun settle(
        obligationId: Id,
        personId: Id,
        amountMinor: Halalas,
        transactionId: Id? = null,
        requestId: String? = null,
    ): Settlement {
        if (requestId != null && !REQUEST_ID.matches(requestId)) throw IllegalStateException("معرّف طلب التسوية غير سليم")
        return deps.settlementWriter.settle(
            Settlement(
                id = if (requestId != null) "stl-$obligationId-$requestId" else deps.ids.next("stl"),
                transactionId = transactionId ?: "",
                obligationId = obligationId,
                amountMinor = amountMinor,
            ),
            personId,
        )
    }

    /**
     * دين قديم من غير عملية — OVERRIDES §27. الالتزام بـ`originTransactionId = null`
     * وبيتسدد عادي، و**مفيش عملية وهمية**: مش بيدخل المصروف ولا الدخل ولا رصيد الكاش.
     */
    suspend fun addOpeningDebt(personId: Id, kind: ObligationKind, amountMinor: Halalas): Obligation {
        if (amountMinor <= 0 || amountMinor > MAX_SAFE_HALALAS) throw IllegalStateException("المبلغ لازم يكون أكبر من صفر")
        if (deps.people.listAll().none { it.id == personId }) throw IllegalStateException("الشخص ده مش موجود")
        val obligation = Obligation(
            id = deps.ids.next("obl"),
            personId = personId,
            originTransactionId = null,
            kind = kind,
            originalMinor = amountMinor,
            currency = Currency.SAR,
        )
        deps.obligations.saveMany(listOf(obligation))
        return obligation
    }
}
